typescript/schedules.md

```diff
@@ -0,0 +1,1257 @@
+   * Window could lead to missed Actions. (But you can always {@link ScheduleHandle.backfill}.)
+   *
+   * @default 1 minute
+   */
+  catchupWindow: Milliseconds;
+
+  /**
+   * When an Action times out or reaches the end of its Retry Policy, {@link pause}.
+   *
+   * With {@link ScheduleOverlapPolicy.AllowAll}, this pause might not apply to the next Action, because the next Action
+   * might have already started previous to the failed one finishing. Pausing applies only to Actions that are scheduled
+   * to start after the failed one finishes.
+   *
+   * @default false
+   */
+  pauseOnFailure: boolean;
+
+  /**
+   * Informative human-readable message with contextual notes, e.g. the reason
+   * a Schedule is paused. The system may overwrite this message on certain
+   * conditions, e.g. when pause-on-failure happens.
+   */
+  note?: string;
+
+  /**
+   * Is currently paused.
+   *
+   * @default false
+   */
+  paused: boolean;
+
+  /** Whether the number of Actions to take is limited. */
+  actionsAreLimited: boolean;
+
+  /**
+   * Limit the number of Actions to take.
+   *
+   * This number is decremented after each Action is taken, and Actions are not
+   * taken when the number is `0` (unless {@link ScheduleHandle.trigger} is called).
+   *
+   * @default unlimited
+   */
+  remainingActions?: NonNegativeInteger;
+}
+
+export interface ScheduleListPage {
+  schedules: Schedule[];
+
+  /**
+   * Token to use to when calling {@link ScheduleClient.listByPage}.
+   *
+   * If `undefined`, there are no more Schedules.
+   */
+  nextPageToken?: string;
+}
+
+export type WorkflowExecution = Required<temporal.api.common.v1.IWorkflowExecution>;
+
+export interface ScheduleInfo {
+  /** Number of Actions taken so far. */
+  numActionsTaken: number;
+  // TODO or numberOfActionsTaken etc?
+
+  /** Number of times a scheduled Action was skipped due to missing the catchup window. */
+  numActionsMissedCatchupWindow: number;
+
+  /** Number of Actions skipped due to overlap. */
+  numActionsSkippedOverlap: number;
+
+  /**
+   * Currently-running workflows started by this schedule. (There might be
+   * more than one if the overlap policy allows overlaps.)
+   * Note that the run_ids in here are the original execution run ids as
+   * started by the schedule. If the workflows retried, did continue-as-new,
+   * or were reset, they might still be running but with a different run_id.
+   */
+  runningWorkflows: WorkflowExecution[];
+
+  /**
+   * Most recent 10 Actions started (including manual triggers).
+   *
+   * Sorted from older start time to newer.
+   */
+  recentActions: ScheduleAction[];
+
+  /** Next 10 scheduled Action times */
+  nextActionTimes: Date[];
+
+  createdAt: Date;
+  lastUpdatedAt: Date;
+
+  isValid: boolean;
+
+  /** Error for invalid schedule. If this is present, no actions will be taken. */
+  invalidScheduleError?: string;
```

Comment:
This and `isValid` seem a bit weird here, when would a schedule reach an invalid state @dnr?

Reply:
This is what you get when you put strings in a calendar spec that can't be parsed. Or a negative interval, or something like that. We can't do those validations synchronously because of the workflow+signal-based implementation. Even if we could (after sync update, maybe), I'd like to reserve an async error channel in case there are errors that can't be detected until we attempt to take an action, for example. (To be clear, currently the only thing is calendar spec validation errors, the others are ideas.)

Reply:
There are different ways to implement sync update without the upcoming sync update feature.
Wondering which other errors you have in mind.

Reply:
Sure but why add that complexity when sync update is coming soon?

I was thinking things like validation errors on the start workflow params, size limits and things. Some of these are checked at schedule creation time but maybe not all. Or dynamic config could change in between and valid params can become invalid. These don't necessarily have to be reported as errors here, of course, it could just keep trying.

Reply:
I assumed we can validate the schedule content up front, not sure which dynamic config values you're referring to and if we'd want to invalidate all schedules when they change.

The reason to add this complexity would be so we can give users the best possible API, we can optimize the backend later when sync update is implemented without changing the experience.
Checking the error asynchronously is suboptimal IMHO.

Reply:
The size limits on workflow id length, task queue name length, input size, etc. are all dynamic config, and could theoretically turn a valid schedule invalid or vice versa after-the-fact. It's a little contrived, I admit, but we should decide what makes most sense in those situations.

We have a channel to return sync errors (rpc return values) and a channel for async errors (this field). Right now some errors are reported sync, some async. Over time I plan to move more to sync, whether that's done with sync update or another mechanism.

So the two questions for now are:

1. Will we ever be able to report _all_ errors synchronously, so that we could get rid of this field? I'm not sure. For workflow start params and schedule spec, yeah probably, but then there may be other action types in the future.

2. Do we have time move all existing errors to sync-style before we want to release this (whether with sync update or another mechanism)? I agree it would be good, but I don't think I have time to do that. Help is appreciated if one of you has bandwidth :)

Reply:
I can totally understand why this isn't the most urgent thing to change, we can always add more sync errors and deprecate this field. My fear is that if we don't do this now it'll be in the backlog forever.

Reply:
Field will be removed before Schedules goes GA. Create and update will return sync errors